### PR TITLE
Tombstone retention

### PR DESCRIPTION
**Pull request by kbatuigas**
## Description

Support for topic level `delete.retention.ms` and cluster level `tombstone_retention_ms` properties, which define the retention limit for tombstone records in compacted topics.

TODO:
- A separate PR will add the `tombstone_retention_ms` reference entry: https://github.com/redpanda-data/docs/pull/847
- We will also have to update https://docs.redpanda.com/current/develop/kafka-clients/#unsupported-kafka-features when this is available.

Resolves https://github.com/redpanda-data/documentation-private/issues/2600
Review deadline: Nov 12

## Page previews

24.3 Beta > [Compaction settings > Tombstone record removal](https://deploy-preview-829--redpanda-docs-preview.netlify.app/24.3/manage/cluster-maintenance/compaction-settings/#tombstone-record-removal)
Topic properties > [delete.retention.ms](https://deploy-preview-829--redpanda-docs-preview.netlify.app/24.3/reference/properties/topic-properties/#deleteretentionms)
[What's new](https://deploy-preview-829--redpanda-docs-preview.netlify.app/24.3/get-started/whats-new/#tombstone-removal)

## Checks

- [ ] New feature
- [ ] Content gap
- [ ] Support Follow-up
- [ ] Small fix (typos, links, copyedits, etc)

### CODE REVIEW

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: Can we provide an explanation for why?

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: This is not entirely accurate to what a tombstone record is.

A tombstone record is a record _explicitly_ produced by a Kafka client with a given key and the value `null`. 
Tombstone records are not generated by a broker or by regularly scheduled garbage collection.
It itself is the "signal" that records with the same key prior to it in the log should be deleted.

Tombstones are removed either by:

1. Typical garbage collection in a `compact,delete` cleanup policy topic per `delete` retention configuration.
2. In a `compact` only topic, tombstones must be "seen" at least twice by the compaction scheduler before they are removed. This is in order to avoid races between consumers and tombstone removal. The value `delete.retention.ms` sets the retention time for a tombstone, and therefore the bound on the time a consumer has for reading from the log before they see an incomplete view (due to tombstone removal). See Confluent's description of `delete.retention.ms` [here](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#delete-retention-ms), and another [helpful resource here](https://www.naleid.com/2023/07/30/understanding-kafka-compaction.html#how-do-i-delete-all-events-sharing-a-key-on-a-compacted-topic).

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: >When all messages for a given key are deleted, the topic partition keeps one message, called a tombstone, that contains the key, and a null value. 

See other comment about why this is not accurate to what a tombstone record is- this is almost a correct view, but is misleading. Tombstones are not generated by regularly scheduled garbage collection, they are produced by Kafka clients.

>A tombstone indicates to clients that the messages have been deleted.

No, it is a request by the client that previous records with the same key _should_ be deleted. Until compaction occurs, a tombstone record and a previous record(s) with the same key may exist and be visible to consumers simultaneously.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: Somewhat misleading, this is closer to "un-setting" the topic level property so that it falls through to the cluster default.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: I don't think we should- tiered storage tombstone removal is a very large undertaking/project that we don't explicitly have as a current focus.

Perhaps phrasing it as "Redpanda does not currently remove tombstone records for compacted topics with Tiered Storage enabled" would be best, cc: @mattschumpert 

This line may also not be needed since you mention that configuration properties cannot be enabled in the same note.

Comment (mattschumpert) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: @WillemKauf is it more accurate to say ' we do not currently remove tombstone records from log segments in compacted topics after segments have been uploaded to cloud storage.  tombstone deletion only occurs in topic data while the data resides in local storage' ?

if that's not true, your suggestion is fine.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: @WillemKauf two questions here - 

- Is there a possible configuration where the compaction schedule (`log_compaction_interval`?) conflicts or "interferes" with what's set for tombstone retention? For example, tombstone retention limit is set too low that the compaction scheduler is not going to see it at least twice?
- I'd love to have the two sentences in this paragraph be better tied together... Is it correct to say that: because the tombstone retention sets the time bound for consumers, tombstones can only be deleted after they are seen twice by the compaction scheduler?

File: modules/reference/pages/properties/topic-properties.adoc
Comment (kbatuigas) [inline modules/reference/pages/properties/topic-properties.adoc:561]: @WillemKauf I'm asking @Deflaimun to pull the cluster config reference description so we can add it to the docs, but afaict there is no default value for `tombstone_retention_ms` either. Is there a value that we recommend users set? Or perhaps we can provide some guidance or scenarios users can look out for so they can set a tombstone retention policy that works well for them?

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: > For example, tombstone retention limit is set too low that the compaction scheduler is not going to see it at least twice?

No, because tombstones being marked as "seen" for the first time as well as the act of tombstone removal are both actions that occur during the compaction routine. There is no possible conflict here. 

> because the tombstone retention sets the time bound for consumers, tombstones can only be deleted after they are seen twice by the compaction scheduler?

I'm starting to think that "tombstones must be seen at least twice by the compaction scheduler before they are removed" is too much of an implementation detail to include here in the docs.

I think we could opt for the more succinct:

The retention time of tombstones in a compacted topic is controlled by the topic property `delete.retention.ms`, or the cluster property `tombstone_retention_ms`. The value used sets the time bound that a consumer has in order to see a complete view of the log with tombstones present before their removal.

Or something there-abouts.

File: modules/reference/pages/properties/topic-properties.adoc
Comment (WillemKauf) [reply modules/reference/pages/properties/topic-properties.adoc:561]: That's right, both are set to `null` by default in order to reduce configuration headaches with tiered storage property conflicts.

Confluent/Kafka default this value to one day (24 hours, `86400000` ms).

Guidance/scenarios:

* If more frequent cleanup of tombstone records is important for space management, this value may be reduced.
* If obtaining a complete snapshot of the log, including tombstone records, is important to your consumers, this value should be set such that consumers have enough time for their reads to complete before tombstones are removed.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (kbatuigas) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: @WillemKauf just to double check, it's the topic retention limit here and not tombstone?

Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: The point we are trying to communicate here is that in `compact,delete` configured topics, tombstones can be removed by the compaction process per the tombstone limits (as we have described in the `compact` process with `delete.retention.ms`), but standard garbage collection would also apply to the topic (e.g if the log were to be garbage collected up to an offset 10, all records with offsets < 10 would be removed, including tombstones records in that range).

File: modules/get-started/pages/whats-new.adoc
Comment (WillemKauf) [inline modules/get-started/pages/whats-new.adoc]: >by setting `delete.retention.ms` on the topic level, or `tombstone_retention_ms` on the cluster level.

Reword? 

> by setting `delete.retention.ms` **at** the topic level, or `tombstone_retention_ms` **at** the cluster level

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: > When a topic's messages should be deleted

"Messages" and "records" are used synonymously in redpanda docs, but is there a preferred term? I.e [rpk topic consume](https://docs.redpanda.com/redpanda-cloud/reference/rpk/rpk-topic/rpk-topic-consume/) mostly refers to "records" (including when it refers to tombstones)

Also, the way this first sentence is phrased makes it sound like a single tombstone record can be responsible for the deletion of many "messages" in a topic. Would rephrase to something like 

> Compaction can also enable the deletion of existing records through the use of tombstones. For example, as data is deleted from a source system, clients produce a tombstone record to the message log. A tombstone contains a key and the value `null`. Tombstones signal to brokers and consumers that records with the same key prior to it in the log should be deleted.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: Taking inspiration from [kafka docs (see point 4)](https://docs.confluent.io/kafka/design/log_compaction.html#compaction-guarantees) once again, we could add a clarifying sentence similar to the following here:

> If obtaining a complete snapshot of the log, including tombstone records, is important to your consumers, set the tombstone retention value such that consumers have enough time for their reads to complete before tombstones are removed. Consumers may not see tombstone records if their reads take longer than `delete.retention.ms` to reach the head of the log.

**Review by WillemKauf — COMMENTED** [review]
It's looking good! A small number of comments on my end.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (kbatuigas) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: @WillemKauf is the last sentence in the paragraph still accurate and relevant? i.e. Can we say "If obtaining a complete snapshot of the log, including tombstone records, is important to your consumers... Consumers may not see tombstone records if their reads take longer than delete.retention.ms to reach the head of the log. _On the other hand, if managing local disk space is a higher priority_..."?

Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: I think something like the following might flow better?

> On the other hand, if more frequent cleanup of tombstone records is important for your workloads and space management, consider setting the tombstone retention to a lower value. For example, less than 24 hours (86400000 ms).

I don't have a great sentence yet, but a statement illustrating the trade-off between tombstone visibility for consumers and space management is still relevant and needed here.

**Comment by Deflaimun** [comment]
@kbatuigas feel free to merge this one when you merge yours 
https://github.com/redpanda-data/docs/pull/847

**Review by mattschumpert — APPROVED** [review]

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: No, there is currently a hard restriction on tombstone removal being enabled along with tiered storage properties on a topic. No tombstones are removed for any TS enabled topics.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: > Consequently, if you set tombstone retention limits that are too low, it could cause more data inconsistencies.

"Data inconsistencies" sounds a bit too scary- consumers not seeing tombstones isn't exactly an "inconsistency".

I think you illustrated the trade-off well enough in the paragraph above on L77, I would recommend removing this sentence or rewording it.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (Deflaimun) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: maybe we can add somewhere here what a tombstone is

File: modules/reference/pages/properties/topic-properties.adoc
Comment (Deflaimun) [inline modules/reference/pages/properties/topic-properties.adoc]: I would remove (in ms) from the description and add a new field "*Units:* milliseconds " like we have on cluster properties.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (Deflaimun) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc:41]: can I enable delete.retention.ms topic property if I have the Tiered Storage cluster properties?

**Review by Deflaimun — APPROVED** [review]
lgtm. Check comments before merging. If you need to address them later, consider creating a new ticket

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc:41]: Yes. Cluster properties are coupled and topic properties are coupled but the `delete.retention.ms` topic property is not coupled to the cluster properties.

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (WillemKauf) [reply modules/manage/pages/cluster-maintenance/compaction-settings.adoc]: Can link to other article in compaction which describes it

**Review by WillemKauf — APPROVED** [review]

File: modules/manage/pages/cluster-maintenance/compaction-settings.adoc
Comment (Feediver1) [inline modules/manage/pages/cluster-maintenance/compaction-settings.adoc]:
```suggestion
This diagram illustrates a compacted topic. Imagine a remote sensor network that uses image recognition to track appearances of red pandas in a geographic area. The sensor network employs special devices that send records to a topic when they detect one. You might enable compaction to reduce topic storage while still maintaining a record in the topic of the last time each device saw a red panda, perhaps to see if they stop frequenting a given area. The left side of the diagram shows all records sent across the topic. The right side illustrates the results of compaction; older records for certain keys are deleted from the log.
```

**Review by Feediver1 — APPROVED** [review]

**Pull request merged**
